torch/jit/_builtins.py

```diff
@@ -139,3 +139,49 @@ def _register_builtin(fn, op):
 
 def _find_builtin(fn):
     return _get_builtin_table().get(id(fn))
+
+
+# These are Python globals that have special sugaring in the compiler. See
+# [python globals] in the compiler for details. Adding a builtin global here
+# indicates that when resolving, we should skip these so they dispatch to
+# the correct sugared representation
+compiler_builtins = None
+def _is_global_builtin(maybe_builtin):
```

Comment:
Right now if someone does

```python
def list(x):
    return x + 1

@torch.jit.script
def fn(x):
    return list(x)
```

It will use `aten::list` instead of compiling the actual function since our resolution is not correct. I ran into this when I added `AssertionError` and `RuntimeError` as globals in the compiler (we need those to support `RuntimeError`s inside `cu.define` which is used in a couple places already), so this says that if we have a builtin Python function, don't resolve it with the resolver if it is the actual builtin.

This does add an extra step, but whenever someone adds a new compiler builtin it won't work until they add it here (since it will just resolve to the Python version and get put in as a Python call), so I think this is a pretty safe change

Reply:
I don't think we need to support `RuntimeError` inside `cu.define`. We don't generally support everything in string parsing that we do normally. See: https://github.com/pytorch/pytorch/issues/31391#issuecomment-567206434

Reply:
We already have code that uses it (I forget where exactly though) and the alternative to have a raise inside of a string is to do something like `raise "exception"`, which goes against the whole "its just a subset of Python" thing. It also makes things like `AssertionError` easier to implement since it just de-sugars `assert cond, msg` to an if statement with a `raise AssertionError(msg)`, but the compiler needs to know how to compile an exception to do that

Reply:
By code that uses it do you mean `checkScript` tests. There are 63 builtin exceptions. Why should we support `AssertionError` and `RuntimeError` but not `ValueError` or any of the other ones ? 

Other things we don't support with compilation unit: `torch.dtype`, list comprehensions, class types, etc. All of our existing files are serialized as `raise "exception"` so this is BC we have to support. Given that we need to support this it should be not very hard to support `assert cond, msg`.

Reply:
Going through it again I couldn't find any case of `raise RuntimeError` in a `.define`, so I'm not sure what was going on there, so we don't need `RuntimeError` as a compiler builtin.

But that doesn't change the general need for this since the compiler is incorrect today since it doesn't respect shadowing Python builtins. It's probably a rare case but I don't think this adds any huge tech debt since there's no way for this list to fall out of date with the one in the compiler.

Where are exceptions serialized as `raise "error"`? I can't find any reference to that in `python_print.cpp` and trying it out they get serialized as `ops.prim.RaiseException("error")`

Reply:
Okay, looks like i was wrong on how its serialized. 

Shadowing of Python builtins is a real problem but it would make this easier to review if that PR was separated from this with this PR. Potentially via ghstack. The only values that we need as compiler builtins are ones that we need to support serialized models. "list" only ever gets serialized as `torch.ops.list`, so it doesn't need to be a compiler builtin. It could instead be registered in `_builtins` as `Iid(list), aten::list)`.

Reply:
That and there are a few other issues with these builtins (e.g. #34851) that we could do as a follow up. Agreed it probably should have been a stack but since this depends on the builtin resolution for the changes to asserts the changes are pretty tangled up now .

Reply:
You're making a lot of changes to both compiler internals and one of the more tricky parts of the code base (sugared values / python sugared values), i think it's a reasonable ask to separate to two PRs.